### PR TITLE
[GL] Android emulator support

### PR DESCRIPTION
**Pull request by Gordon-F**
PR checklist:
- [x] tested examples with the following backends: OpenGL ES

Android emulator doesn't support EGL_KHR_surfaceless_context.

### CODE REVIEW

File: src/backend/gl/src/window/egl.rs [hunk 1]
```diff
@@ -596,9 +613,6 @@ impl Surface {
         gl.bind_framebuffer(glow::READ_FRAMEBUFFER, None);
 
         self.egl.swap_buffers(self.display, self.raw).unwrap();
-        self.egl
```
Comment (Gordon-F) [inline src/backend/gl/src/window/egl.rs:599]: I'm not sure about these changes. I don't know why we should make the current context without surface every frame?

File: src/backend/gl/src/window/egl.rs [hunk 2]
```diff
@@ -114,11 +115,9 @@ impl Inner {
             return Err(hal::UnsupportedBackend);
         }
         let display_ext_list = display_extensions.split_whitespace().collect::<Vec<_>>();
-        let required_display_extensions = ["EGL_KHR_create_context", "EGL_KHR_surfaceless_context"];
-        for required_ext in required_display_extensions.iter() {
-            if !display_ext_list.contains(required_ext) {
-                log::warn!("{} is not present", required_ext);
-            }
+        if !display_ext_list.contains(&"EGL_KHR_create_context") {
```
Comment (kvark) [inline src/backend/gl/src/window/egl.rs:118]: do we actually require this? My impression was that we could get away even without it

File: src/backend/gl/src/window/egl.rs [hunk 3]
```diff
@@ -285,10 +285,27 @@ impl hal::Instance<crate::Backend> for Instance {
...
+        if inner.supports_surfaceless_context {
+            inner
+                .egl
+                .make_current(inner.display, None, None, Some(inner.context))
+                .unwrap();
+        } else {
+            let attributes = [egl::WIDTH, 1, egl::HEIGHT, 1, egl::NONE];
+            let surface = inner
+                .egl
+                .create_pbuffer_surface(inner.display, inner.config, &attributes)
```
Comment (kvark) [inline src/backend/gl/src/window/egl.rs:297]: Why do we need a PBuffer surface? We request the surface-less context only because we can't use the provided surfaces efficiently. If the context has to have them, it's ok, we shouldn't need to do anything, still

File: src/backend/gl/src/window/egl.rs (hunk 1 above)
Comment (kvark) [reply src/backend/gl/src/window/egl.rs:599]: It's just returning everything back to what it is before the `present()` is called. We could technically leave the surface bound, but it could introduce accidental code errors when something is rendering to the default framebuffer, and we aren't catching a GL error on that (instead corrupting one of the swapchains). Note that the `self.context` can work with different surfaces, not just one.

**Review by kvark — COMMENTED** [review]
Thank you for the PR!

File: src/backend/gl/src/window/egl.rs (hunk 3 above)
Comment (Gordon-F) [reply src/backend/gl/src/window/egl.rs:297]: You mean `create_window_surface` and destroy after `make_current` ?

File: src/backend/gl/src/window/egl.rs (hunk 1 above)
Comment (Gordon-F) [reply src/backend/gl/src/window/egl.rs:599]: Any ideas how we can get around this limitation?

File: src/backend/gl/src/window/egl.rs (hunk 3 above)
Comment (kvark) [reply src/backend/gl/src/window/egl.rs:297]: Sorry, I don't understand. Let me try again.

So this is creating a PBuffer because otherwise we can't call `make_current`, correct?
In this case, could we perhaps make it a 1x1 buffer instead of having the full width/height?

Also, we need to keep it around, so that we can call `make_current` later again (see the other comment we are discussing)

File: src/backend/gl/src/window/egl.rs (hunk 1 above)
Comment (kvark) [reply src/backend/gl/src/window/egl.rs:599]: Yep, we can keep the `PBuffer` around and use it here when calling `make_current`

File: src/backend/gl/src/window/egl.rs (hunk 3 above)
Comment (Gordon-F) [reply src/backend/gl/src/window/egl.rs]: > So this is creating a PBuffer because otherwise we can't call make_current, correct?

Correct.

> In this case, could we perhaps make it a 1x1 buffer instead of having the full width/height?

```rust
let attributes = [egl::WIDTH, 1, egl::HEIGHT, 1, egl::NONE];
```

It should be 1x1.

**Comment by Gordon-F** [comment]
@kvark I think I fixed all suggestions. Thank you for reviewing!

File: src/backend/gl/src/window/egl.rs [hunk 4]
```diff
@@ -285,10 +298,23 @@ impl hal::Instance<crate::Backend> for Instance {
 
     fn enumerate_adapters(&self) -> Vec<hal::adapter::Adapter<crate::Backend>> {
         let inner = self.inner.lock();
-        inner
-            .egl
-            .make_current(inner.display, None, None, Some(inner.context))
-            .unwrap();
+        if let Some(pbuffer) = inner.pbuffer {
+            inner
```
Comment (kvark) [inline src/backend/gl/src/window/egl.rs:302]: do we actually need this `if`? couldn't we just pass `make_current(inner.display, inner.pbuffer, inner.pbuffer, ..)` right away?

File: src/backend/gl/src/window/egl.rs [hunk 5]
```diff
@@ -596,9 +625,21 @@ impl Surface {
         gl.bind_framebuffer(glow::READ_FRAMEBUFFER, None);
 
         self.egl.swap_buffers(self.display, self.raw).unwrap();
-        self.egl
-            .make_current(self.display, None, None, Some(self.context))
-            .unwrap();
+
+        if let Some(pbuffer) = self.pbuffer {
```
Comment (kvark) [inline src/backend/gl/src/window/egl.rs:629]: same here, perhaps we can remove the `if`?

**Review by kvark — COMMENTED** [review]
Looks wonderful, just one small nit left!

File: src/backend/gl/src/window/egl.rs (hunk 3 above)
Comment (kvark) [reply src/backend/gl/src/window/egl.rs]: Sorry, I just missed it. The code was good!

File: src/backend/gl/src/window/egl.rs (hunk 4 above)
Comment (Gordon-F) [reply src/backend/gl/src/window/egl.rs]: Removed `if` and checked on the emulator and device. Everything is working fine.

**Comment by kvark** [comment]
Sweet!
Bors r+
